### PR TITLE
Custom transfer amount to hydroponic trays

### PR DESCRIPTION
**Pull request by Brian-Archambault**
## Set transfer amount to hydroponic trays 

**Screenshots**

transfer

**Changelog**

:cl:
- Can fill hydroponic trays with the transfer amount set on container (bucket) instead of base 5 per click (within 5-50 units capped by how much solution is left/room in tray)

Open to feedback since this is my first PR and it might change botany a bit

### CODE REVIEW

**Comment by ElectroJr** [comment]
This is a good issue to try fix for a first PR. While the current fix you have definitely works I would go about fixing it in a different way, thought I think its still easily do-able for a first time PR. If you have any questions feel free to ask me on discord.

Firstly, If you give a component min/max receive amount like you did here, you should generally make it a data-field so that it can be specified in yaml files, rather than having some hardcoded amount. Also, I don't think having a minimum receive amount makes sense.  If you tried to use a dropper to add 1u of reagent, it would override the desired transfer amount and just take more reagent than it should.

Secondly, it turns out that hydroponics trays also have a `RefillableSolutionComponent`, which already provides solution transfer interactions (see `SolutionTransferComponent.cs`), and these interactions already respect transfer amounts. It just so happens that the plant-holder interaction always occurs before that one, so the `RefillableSolutionComponent` goes completely ignored.

I'm not sure why it is currently like that, (probably just old code), but IMO `PlantHolderComponent` just shouldn't have that transfer logic at all. It should just rely on `RefillableSolutionComponent` instead of duplicating that interaction, especially as the interaction here is very general not specific to hydroponic trays. The one unique feature appears to be that if the solution-source is a spray bottle, it makes a sound & uses slightly different popup text, so maybe that can stay.

So I would suggest that you update the PR to make the max-receive-quantity a data-field of the `RefillableSolutionComponent`, and then remove all the non-spray bottle fluid transfer logic from `PlantHolderComponent` in favour of just using the interactions in `SolutionTransferComponent.cs` 

Optionally, you could also generalize `DrainableSolutionComponent` so that it has data-fields for sound and popup-text and then make the spray bottle transfer interaction use those. That way the spray bottle interaction could also just be generalized and removed from `PlantHolderComponent`, and they would then always make a sound when adding reagents to something, whether that be hydroponic trays or beakers. Though maybe that should just be a separate PR to keep things simple.

**Comment by Brian-Archambault** [comment]
Thanks for the feedback. I added a maxrefill that caps the amount in solution transfer for trays and a cappedfill so that the rest of the yaml files don't need to be updated unless they also have a capped receiving amount. 

I made the botany solution code only apply to sprays for now so I don't accidently break them in this PR but I think they could also be moved into solution transfer later.

File: Content.Server/Botany/Components/PlantHolderComponent.cs [hunk 1]
```diff
@@ -728,23 +729,21 @@ async Task<bool> IInteractUsing.InteractUsing(InteractUsingEventArgs eventArgs)
...
             var solutionSystem = EntitySystem.Get<SolutionContainerSystem>();
             if (solutionSystem.TryGetDrainableSolution(usingItem, out var solution)
-                && solutionSystem.TryGetSolution(Owner, SoilSolutionName, out var targetSolution))
+                && solutionSystem.TryGetSolution(Owner, SoilSolutionName, out var targetSolution) && _entMan.TryGetComponent(usingItem, out SprayComponent? spray))
             {
-                var amount = FixedPoint2.New(5);
-                var sprayed = false;
+                var amount = FixedPoint2.New(1);
+
+                var sprayed = true;
```
Comment (ElectroJr) [inline Content.Server/Botany/Components/PlantHolderComponent.cs:736]: Given that sprayed is always true now, you can just remove it.

File: Content.Server/Chemistry/Components/SolutionManager/RefillableSolutionComponent.cs [hunk 2]
```diff
@@ -21,5 +22,21 @@ public class RefillableSolutionComponent : Component
...
+
+        /// <summary>
+        /// Does this item have a max refill set?
+        /// </summary>
+        [DataField("cappedfill")]
+        [ViewVariables(VVAccess.ReadWrite)]
+        public bool CappedFill { get; set; } = false;
+
+
+
```
Comment (ElectroJr) [inline Content.Server/Chemistry/Components/SolutionManager/RefillableSolutionComponent.cs:40]: Maybe just make MaxRefill nullable & default to null.
```suggestion
        public FixedPoint2? MaxRefill { get; set; } = null;
```

File: Content.Server/Chemistry/Components/SolutionManager/RefillableSolutionComponent.cs [hunk 3]
```diff
@@ -21,5 +22,21 @@ public class RefillableSolutionComponent : Component
         [DataField("solution")]
         public string Solution { get; set; } = "default";
 
+        /// <summary>
+        /// The maximum amount that can be transferred to the solution at once
+        /// </summary>
+        [DataField("maxrefill")]
```
Comment (ElectroJr) [inline Content.Server/Chemistry/Components/SolutionManager/RefillableSolutionComponent.cs:28]: use camelCase (i.e., "maxRefill") for data-fields. Some older bits of code don't always do that.

File: Content.Server/Botany/Components/PlantHolderComponent.cs (hunk 1 above)
Comment (Brian-Archambault) [reply Content.Server/Botany/Components/PlantHolderComponent.cs]: Yeah that makes sense, I changed the popup message to just be for sprays and removed sprayed

File: Content.Server/Chemistry/Components/SolutionManager/RefillableSolutionComponent.cs (hunk 2 above)
Comment (Brian-Archambault) [reply Content.Server/Chemistry/Components/SolutionManager/RefillableSolutionComponent.cs]: I don't have a lot of experience with nullability but I'm assuming it's okay to cast back after I check if it's null? It seems to work when I test it anyway.

File: Content.Server/Chemistry/Components/SolutionManager/RefillableSolutionComponent.cs (hunk 3 above)
Comment (Brian-Archambault) [reply Content.Server/Chemistry/Components/SolutionManager/RefillableSolutionComponent.cs]: my bad, fixed.

**Review by PaulRitter — APPROVED** [review]

File: Content.Server/Chemistry/Components/SolutionManager/RefillableSolutionComponent.cs (hunk 2 above)
Comment (ElectroJr) [reply Content.Server/Chemistry/Components/SolutionManager/RefillableSolutionComponent.cs]: Yeah. Nullables are a bit funny with structs. You can also just use `MaxRefil.Value` after checking if its null (or alternatively, checking `MaxRefill.HasValue`). But cast also works.

**Review by ElectroJr — APPROVED** [review]

**Pull request merged**